components/style/properties/helpers/animated_properties.mako.rs

```diff
@@ -757,6 +766,695 @@ impl Interpolate for LengthOrNone {
+            computed_matrix.m21 = decomposed.matrix.m21;
+            computed_matrix.m22 = decomposed.matrix.m22;
+
+            // Translate matrix.
+            computed_matrix.m41 = decomposed.translate.0 * decomposed.matrix.m11 +
+                                  decomposed.translate.1 * decomposed.matrix.m21;
+            computed_matrix.m42 = decomposed.translate.0 * decomposed.matrix.m12 +
+                                  decomposed.translate.1 * decomposed.matrix.m22;
+
+            // Rotate matrix.
+            let angle = decomposed.angle.to_radians();
+            let cos_angle = angle.cos();
+            let sin_angle = angle.sin();
+
+            let mut rotate_matrix = ComputedMatrix::identity();
+            rotate_matrix.m11 = cos_angle;
+            rotate_matrix.m12 = sin_angle;
+            rotate_matrix.m21 = -sin_angle;
+            rotate_matrix.m22 = cos_angle;
+
+            // Multiplication of computed_matrix and rotate_matrix
+            computed_matrix = multiply(computed_matrix, rotate_matrix);
+
+            // Scale matrix.
+            computed_matrix.m11 *= decomposed.scale.0;
+            computed_matrix.m12 *= decomposed.scale.0;
+            computed_matrix.m21 *= decomposed.scale.1;
+            computed_matrix.m22 *= decomposed.scale.1;
+            computed_matrix
+        }
+    }
+
+    #[derive(Clone, Copy, Debug)]
+    #[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+    pub struct Translate3D(f32, f32, f32);
+
+    #[derive(Clone, Copy, Debug)]
+    #[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+    pub struct Scale3D(f32, f32, f32);
+
+    #[derive(Clone, Copy, Debug)]
+    #[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+    pub struct Skew(f32, f32, f32);
+
+    #[derive(Clone, Copy, Debug)]
+    #[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+    pub struct Perspective(f32, f32, f32, f32);
+
+    #[derive(Clone, Copy, Debug)]
+    #[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+    pub struct Quaternion(f32, f32, f32, f32);
+
+    #[derive(Clone, Copy, Debug)]
+    #[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+    pub struct MatrixDecomposed3D {
+        pub translate: Translate3D,
+        pub scale: Scale3D,
+        pub skew: Skew,
+        pub perspective: Perspective,
+        pub quaternion: Quaternion,
+    }
+
+    impl From<ComputedMatrix> for MatrixDecomposed3D {
+        /// Decompose a 3D matrix.
+        /// https://drafts.csswg.org/css-transforms/#decomposing-a-3d-matrix
+        fn from(matrix: ComputedMatrix) -> MatrixDecomposed3D {
+            let mut matrix = matrix;
+            // Normalize the matrix.
+            if matrix.m44 == 0.0 {
+                // TODO(canaltinova): return false here but from tait doesn't return Option or Result
+                // Find a way to do this.
+            }
+
+            % for i in range(1, 5):
+                % for j in range(1, 5):
+                    matrix.m${i}${j} /= matrix.m44;
```

Comment:
While the current ordering is fine, might be better to bind m44 to a variable called "scalingfactor" and divide by that so that it doesn't get accidentally mutated.